wasatch/SPIDevice.py

```diff
@@ -0,0 +1,331 @@
+            if reading.spectrum == False:
+                return False
+        except usb.USBError:
+            self.failure_count += 1
+            log.error(f"SPI Device: encountered USB error in reading for device {self.device}")
+
+        if not reading.failure:
+            if averaging_enabled:
+                if self.sum_count == 0:
+                    self.summed_spectra = [float(i) for i in reading.spectrum]
+                else:
+                    log.debug("device.take_one_averaged_reading: summing spectra")
+                    for i in range(len(self.summed_spectra)):
+                        self.summed_spectra[i] += reading.spectrum[i]
+                self.sum_count += 1
+                log.debug("device.take_one_averaged_reading: summed_spectra : %s ...", self.summed_spectra[0:9])
+
+        if self.settings.eeprom.bin_2x2:
+            # perform the 2x2 bin software side
+            next_idx_values = reading.spectrum[1:]
+            # average all except the last value, which is just appended as is
+            binned = [(value + next_value)/2 for value, next_value in zip(reading.spectrum[:-1], next_idx_values)]
+            binned.append(reading.spectrum[-1])
+            reading.spectrum = binned
+                
+
+        self.session_reading_count += 1
+        reading.session_count = self.session_reading_count
+        reading.sum_count = self.sum_count
+
+        if averaging_enabled:
+            if self.sum_count >= self.settings.state.scans_to_average:
+                reading.spectrum = [ x / self.sum_count for x in self.summed_spectra ]
+                log.debug("spi device acquire data: averaged_spectrum : %s ...", reading.spectrum[0:9])
+                reading.averaged = True
+
+                # reset for next average
+                self.summed_spectra = None
+                self.sum_count = 0
+        else:
+            # if averaging isn't enabled...then a single reading is the
+            # "averaged" final measurement (check reading.sum_count to confirm)
+            reading.averaged = True
+
+        return SpectrometerResponse(reading)
+
+    def write_eeprom(self) -> SpectrometerResponse:
+        try:
+            self.settings.eeprom.generate_write_buffers()
+        except:
+            log.critical("failed to render EEPROM write buffers", exc_info=1)
+            #self.message_queue("marquee_error", "Failed to write EEPROM")
+            return False
+
+        for page in range(EEPROM.MAX_PAGES):
+            self._EEPROMWritePage(page,self.settings.eeprom.write_buffers[page])
+
+        #self.message_queue("marquee_info", "EEPROM successfully updated")
+        return SpectrometerResponse(True)
+
+    def _EEPROMReadPage(self, page: int) -> list[bytes]:
+        while True:
+            EEPROMPage  = bytearray(74)
+            command     = bytearray(7)
+            command     = [0x3C, 0x00, 0x02, 0xB0, (0x40 + page), 0xFF, 0x3E]
```

Comment:
Maybe in the class documentation at the top of this file add a @see Wasatch Photonics ENG-0150 SPI API